util/timer/src/timer.rs

```diff
@@ -0,0 +1,707 @@
+                    entry.remove();
+                // 'schedule.state_control' was detached one (Def 1).
+                // Don't reschedule since it is Cancelled (Rule 1)
+                // schedule is going to be removed
+                } else {
+                    unreachable!("Rule 1 was violated");
+                }
+            }
+        };
+    }
+
+    fn cancel(&mut self, timer_id: TimerId) -> Result<bool, ScheduleError> {
+        // See Corollary 2.
+        match self.states.entry(timer_id) {
+            Entry::Vacant(_) => Err(ScheduleError::TokenNotScheduled),
+            Entry::Occupied(entry) => {
+                // Detach and cancel it (Rule 1)
+                let state_control = entry.remove();
+                Ok(state_control.cancel())
+            }
+        }
+    }
+
+    fn handle_timeout(&mut self, worker_queue: &WorkerQueue) -> Option<Duration> {
+        loop {
+            let now = Instant::now();
+            match self.heap.peek() {
+                None => return None,
+                Some(Reverse(TimeOrdered(earliest))) if now < earliest.at => return Some(earliest.at - now),
+                _ => { /* lifetime prevents modifying heap from here. */ }
+            }
+            let Reverse(TimeOrdered(timed_out)) = self.heap.pop().expect("It always have an item");
+
+            if timed_out.repeat.is_some() {
+                // timed_out.state_control is re-pushed only after it is popped (Rule 3)
+                self.try_reschedule(timed_out.clone());
+                worker_queue.enqueue(timed_out);
+            } else {
+                let set_queued = match self.states.entry(timed_out.timer_id) {
+                    Entry::Occupied(entry) => {
+                        if Arc::ptr_eq(entry.get(), &timed_out.state_control) {
+                            // 'timed_out.state_control' was attached one. (Def 1)
+                            entry.remove();
+                            timed_out.state_control.try_set_queued()
+                        } else {
+                            false
+                        }
+                    }
+                    _ => false,
+                };
+                // timed_out is anyway removed.
+                if set_queued {
+                    worker_queue.enqueue(timed_out);
+                } else {
+                    // 'timed_out.state_control' was detached one. (Def 1)
+                    // It already be cancelled (Rule 1)
+                    debug_assert!(timed_out.state_control.is_cancelled());
+                }
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+struct TimerId(TimerName, TimerToken);
+
+/*
+  valid state transition:
+    Wait -> Queued -> Timeout/Cancel
+    Wait -> Cancel
+    WaitRepeat -> Cancel
+*/
+#[derive(Eq, PartialEq, Debug)]
+enum ScheduleState {
+    Wait,
+    WaitRepeating,
+    Queued,
+    Timeout,
+    Cancelled,
```

Comment:
What's the relation between a detached state and a canceled state?
Are all detached states canceled?

Reply:
Attached/Detached is based on whether the ```state_control``` is contained in ```self.state```, ```self.heap```, not based on its intrinsic state.
There can be an attached ```state_control``` with ```Cancelled``` state if we issue a handle that contains ````state_control```` (to waiting for the timeout, or canceling through the handle, not by ```TimerApi::cancel```)
Since we don't issue a handle, There's very strong relation between Attached/```Wait```, Detached/```Cancel```, but I think calling it Wait, Cancelled might introduce some confusions.

Reply:
So there are four internal states.
1. Attached and waiting.
2. Attached but canceled.
3. Detached and canceled.
4. None of the above. Canceled or the one-shot timer event fired or Not registered.
Is it right?